packages/datadog-instrumentations/src/vercel-ai.js

```diff
@@ -0,0 +1,70 @@
+'use strict'
+
+const { addHook } = require('./helpers/instrument')
+const tracer = require('../../dd-trace')
+const shimmer = require('../../datadog-shimmer')
+
+const { TracerProvider } = tracer
+const provider = new TracerProvider()
+provider.register()
+
+const { channel } = require('dc-polyfill')
+const toolCreationChannel = channel('dd-trace:vercel-ai:tool')
+
+const TRACED_FUNCTIONS = {
+  generateText: wrapWithTracer,
+  streamText: wrapWithTracer,
+  generateObject: wrapWithTracer,
+  streamObject: wrapWithTracer,
+  embed: wrapWithTracer,
+  embedMany: wrapWithTracer,
+  tool: wrapTool
+}
+
+function wrapWithTracer (fn) {
+  return function () {
+    const options = arguments[0]
+    if (options.experimental_telemetry != null) return fn.apply(this, arguments)
+
+    options.experimental_telemetry = {
+      isEnabled: true,
+      // TODO(sabrenner): need to figure out how a user can configure this tracer
+      // maybe we advise they do this manually?
+      tracer: provider.getTracer()
```

Comment:
Ideally we wouldn't go through translation as that could have unintended side-effects and won't be as flexible. Is there a centralized place in the module that handles tracing? Maybe we could patch something there instead?

Reply:
hey do have a `getTracer` [function](https://github.com/vercel/ai/blob/main/packages/ai/core/telemetry/get-tracer.ts), but as this whole module is bundled it seems into one `dist/index.js` ([ref](https://www.npmjs.com/package/ai?activeTab=code)), it's not actually available on module exports for us to hook into.

Reply:
What I mean is that for example if it expected to receive an object with `startSpan` on it, you could just pass an object created in the instrumentation that would just send a message on a channel like other integrations do instead of passing in a provider. Basically converting OTel to what the library should actually be doing itself 😅 Then the plugin can listen to that as normal without any hacks to make it work with OTel.

For example:

```
tracer: {
  startSpan ( ... ) {
    startCh.publish( ... )
  }
}
```

Reply:
to further that point a bit, the dummy 'span' passed into the function in the instrumentation for `startActiveSpan` doesn't have all of the otel methods, just the ones used by vercel-ai. should they use one not on there, we break. imho any solution there that's like a) subclass the span from a noop otel span kind or b) just have all the methods, points us back to just listening for otel spans in the first place 😅

Reply:
> my main concern is if someone wants to pass in their own dd-trace provider otel tracer (as noted is possible in the PR description), we'll need to still handle those events (or just always override the tracer option, which i don't think is a good solution)

i've pushed up one more change in https://github.com/DataDog/dd-trace-js/pull/5858/commits/407d6dc1aec6e8ebe65559e3adc10558eb30dfc1 that addresses this concern, which seems a lot cleaner. but i guess i still have a question on reusability of otel tracer spans from integrations we want to have. if we add more integrations that have their own otel tracing under the hood, is this a pattern we want to do every time?

Reply:
yeah i think this answers all of my questions as well, but i did implement it the other way as stated. will resolve this for now but lmk if there are points in here i missed addressing in code!